README.md
## 🎭Tor Proxy
Questo progetto verrà utilizzato per fare proxy attraverso tor per evitare che venga tracciato e se dovesse bloccare per le troppe richieste viene riavviato con un ip sempre diverso dal precedente.
### Information general:
> Note: `not require` volume mounted on Docker

### Dependencies
| Services | Required |
| ------ | ------ |
| RabbitMQ | ✅  |

### Variabili globali richiesti:
```sh
example:
    #--- General ---
    ADDRESS_RABBIT: "localhost" #"localhost" [default]
    PORT_RABBIT: 9999 #5672 [default]
    USERNAME_RABBIT: "guest" #"guest" [default]
    PASSWORD_RABBIT: "guest" #"guest" [default]
    EXCHANGE_NAME: "example_exchange" [required]
    QUEUE_RABBIT: "example_queue" #"animemanga-tor-proxy" [default]
    REPLICAS: 4 #15 [default]
    EXPECTED_ADDRESS: #ADDRESS_RABBIT [default]
    START_PORT: 9999 #8000 [default]
    PROXY_PATH: "/path/example" #"proxy.txt" [default]
```

changelog
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Released]
## [1.0.0] - 16-10-2023
### Version Stable

utils/proxies.py
### DOCKER ###
import docker

### CONSTANTS ###
# Docker Settings
DOCKER_PROXY_IMAGE = "dockage/tor-privoxy:latest"

### NASTY GLOBAL VARIABLES (AT LEAST THEY ARE GLOBAL ONLY ON THIS MODULE) ###
# Declaring container map
docker_conatiner_map = {}


# Gets the docker client
def get_docker_client():
    """
    # get_docker_client()
    ## Returns the handle to the docker installation to manage
    """

    # Return docker from env (local installation)
    return docker.from_env()


# Initiation function to initialize all proxies
def initiate_proxies(replicas: int, start_port: int, expected_address: str):
    """
    # initiate_proxies()
    ## Function that initializes all proxies managed by the service
    :rerplicas: number of proxies (containers) that need to be started
    :start_port: start port of the proxies (all generated proxies will have a port equal to start_port + proxy number
    :expected_address: address expected from the requests
    """

    try:
        print(f"Intitiating '{replicas}' [{start_port} - {start_port + (replicas - 1)}] proxies ...")

        # Loop <replicas> times to create all proxies
        for replica_num in range(replicas):
            # Calculate the proxy container port that needs to be assigned
            container_port = start_port + replica_num
            # Format the container name
            container_name = f"animemanga-tor-proxy-{container_port}"
            # Construct the container address (used by the incoming messages to identify the proxy they need to manage)
            container_full_address = f"http://{expected_address}:{container_port}"

            print(f"Intitiating '{container_full_address}'")

            # Start the proxies and save their handles
            docker_conatiner_map[container_full_address] = get_docker_client().containers.run(image=DOCKER_PROXY_IMAGE, detach=True, ports={8118: container_port}, name=container_name)

        print("All proxies are up and running!")
    except Exception as e:
        # Raise and exception appending the original error
        raise Exception(f"initiate_proxies(): cannot initiate '{DOCKER_PROXY_IMAGE}' please check your Docker configuration: " + str(e))


# Restart function to restart a specific proxy
def restart_proxy(proxy: str):
    """
    # restart_proxy()
    ## Function that restarts a proxy given it's address
    :proxy: Full expected address of the proxy http://<address>:<port>
    :return bool: True: the proxy server has been found and restarted - False: the proxy server does not exists
    :raise Exception: if an unexpected error occurs
    """

    try:
        # Get proxy from the map
        proxy_container = docker_conatiner_map.get(proxy)

        # Check if the proxy exists
        if proxy_container:
            # If I get here, it exists and I can restart the proxy server
            proxy_container.restart()
            return True
        else:
            # If I get here, it does not exists so I'll do nothing
            return False
    except Exception as e:
        raise Exception("restart_proxy(): " + str(e))


# Termination function for all proxies
def terminate_proxies():
    """
    # terminate_proxies()
    ## Function that stops all proxies managed by the service
    """

    try:
        print("Terminating proxies ...")

        # Turn off all of the containers for the proxies
        for key, container in docker_conatiner_map.items():
            print(f"Terminating '{key}' ...")

            # Stopping & Removing container
            container.stop()
            container.remove()

        print("All proxies have been terminated!")
    except Exception as e:
        # Raise and exception appending the original error
        raise Exception("terminate_proxies(): an error occurred while terminating and removing proxies, please remove any remaining proxies manually and Check docker your configuration: " + str(e))


# Generates the proxy file
def write_proxy_file(path: str, separator: str = ','):
    """
    # write_proxy_file()
    ## Writes the proxy file at the path indicated
    :path: full path to the proxy file to be written
    """

    # Don't do s*** if theres no path (obviously!!)
    if(len(path) == 0):
        return

    # Loop all keys and join them togheter separating with <separator> character
    proxy_list_str = separator.join([key for key in docker_conatiner_map.keys()])

    # Open file at <path> in write mode (the path MUST be full)
    with open(path, 'w') as proxy_file:
        # Write the proxy file
        proxy_file.write(proxy_list_str)


utils/rabbitmq.py
### RABBITMQ ###
import pika
from pika.exceptions import AMQPConnectionError

### NASTY GLOBAL VARIABLES (AT LEAST THEY ARE GLOBAL ONLY ON THIS MODULE) ###
# Declaring connection and channel
connection = None
channel = None


# Connects to the RabbitMQ server
def connect(host: str, port: int, user: str, passwd: str):
    # In order for the function to see the connection and channel variables
    global connection
    global channel

    try:
        # Establish a connection to RabbitMQ server
        print(f"Connecting to rabbitmq://{host}:{port} ...")

        # Setup of plain credentials
        credentials = pika.PlainCredentials(username=user, password=passwd)

        # Setup connection parameters
        parameters = pika.ConnectionParameters(
            host=host,
            port=port,
            credentials=credentials,
            connection_attempts=10,  # Number of retry attempts if connection fails
            retry_delay=5,           # Delay in seconds between connection retries
            socket_timeout=10,       # Socket operation timeout in seconds
        )

        # Setup of the connection
        connection = pika.BlockingConnection(parameters=parameters)
        # Connecting to the RabbitMQ service
        channel = connection.channel()

        print(f"Connection to rabbitmq://{host}:{port} estabished!")
    except AMQPConnectionError:
        # Raise and exception
        raise Exception("Can't connect to RabbitMQ (check ip/hostname:port or that credentials are correct)")


# Declares an exchange
def exchenage_declare(exchange_name: str, type: str = 'fanout', durable: bool = False):
    # Declare the exchange
    channel.exchange_declare(exchange=exchange_name, exchange_type=type, durable=durable)


# Declares a queue
def queue_declare(queue_name: str):
    # Declaring the message queue
    channel.queue_declare(queue=queue_name)
    print(f"Delcared queue '{queue_name}'")


#  Bind queue to exchange
def bind_queue(exchange_name: str, queue_name: str):
    channel.queue_bind(exchange=exchange_name, queue=queue_name)


# Declare how the messages will be handled by the consumer
def basic_consume(queue: str, on_message_callback, auto_ack: bool):
    channel.basic_consume(queue=queue, on_message_callback=on_message_callback, auto_ack=auto_ack)


# Start consuming messages
def start_consuming():
    try:
        print('Waiting for messages. To exit, press CTRL+C')
        channel.start_consuming()
    except pika.exceptions.ConnectionClosedByBroker as e:
        print("start_consuming() --> Connection lost form RabbitMQ: " + str(e))


# Close the rabbitmq connection
def close_connection():
    try:
        # Closing the RabbitMQ connection
        print('Closing RabbitMQ connection ...')
        channel.stop_consuming()
        connection.close()
    except Exception as e:
        raise Exception("There was and error while closing the connection from RabbitMQ: " + str(e))


app.py
import os
import sys
import json
import signal
from argparse import ArgumentParser

### UTILS ###
from utils import proxies
from utils import rabbitmq

### CONSTANTS ###
# ArgumentParser Settings
PRG_NAME = "animemanga-tor-privoxy"
PRG_DESC = "Anime Manga - tor-privoxy container manager"
PRG_EPIL = f"V1 - {PRG_NAME}"
# ArgumentParser Defaults
DEF_RABBIT_HOST = os.environ.get('ADDRESS_RABBIT', "localhost")
DEF_RABBIT_PORT = os.environ.get('PORT_RABBIT', 5672)
DEF_RABBIT_USER = os.environ.get('USERNAME_RABBIT', "guest")
DEF_RABBIT_PASS = os.environ.get('PASSWORD_RABBIT', "guest")
DEF_RABBIT_EXCH = os.environ.get('EXCHANGE_NAME', "")
DEF_RABBIT_QUEUE = os.environ.get('QUEUE_RABBIT', "animemanga-tor-proxy")
DEF_PROXY_REPLICAS = os.environ.get('REPLICAS', 15)
DEF_EXPECTED_ADDR = os.environ.get('EXPECTED_ADDRESS', DEF_RABBIT_HOST)
DEF_START_PORT = os.environ.get('START_PORT', 8000)
DEF_PROXY_PATH = os.environ.get('PROXY_PATH', "proxy.txt")


### FUNCTIONS ###
# Define a function to handle termination signals
def handle_termination(signal, frame):
    print("Closing gracefully, please wait ...")

    # Terminate all started proxies (if any)
    proxies.terminate_proxies()
    # Close connection with RabbitMQ
    rabbitmq.close_connection()

    # Exit with OK
    sys.exit(0)


# Register the signal handler for SIGINT and SIGTERM (CTLR+C and unexpected interrupts)
signal.signal(signal.SIGINT, handle_termination)
signal.signal(signal.SIGTERM, handle_termination)


# Callback called when messages come in from the queue
def callback(ch, method, properties, body):
    # Parse the message into a Json object
    json_object = json.loads(body.decode('UTF-8'))['message']
    print(f"Received payload: {json_object}")

    # Go trought all possible actions
    if(json_object['action'] == "restart"):  # Action: restart
        # Restart proxy
        if(proxies.restart_proxy(json_object['endpoint'])):
            print(f"Proxy '{json_object['endpoint']}' restarted ...")

            # Manually acknowledge (ack) the message (the proxy has been restarted succesfully)
            ch.basic_ack(delivery_tag=method.delivery_tag)


### ENTRY POINT ###
def main():
    parser = ArgumentParser(PRG_NAME, description=PRG_DESC, epilog=PRG_EPIL)
    parser.add_argument('--rabbit-host', default=DEF_RABBIT_HOST, help="ip/hostname of the RabbitMQ service")
    parser.add_argument('--rabbit-port', default=DEF_RABBIT_PORT, help="port of the RabbitMQ service [default: 5672]")
    parser.add_argument('--rabbit-user', default=DEF_RABBIT_USER, help="username of the RabbitMQ service")
    parser.add_argument('--rabbit-pass', default=DEF_RABBIT_PASS, help="password of the RabbitMQ service")
    parser.add_argument('--exchange-name', default=DEF_RABBIT_EXCH, help="name of the exchange that needs to be created")
    parser.add_argument('--queue-name', default=DEF_RABBIT_QUEUE, help="name of the message queue that needs to be created")
    parser.add_argument('--replicas', default=DEF_PROXY_REPLICAS, type=int, help="the amount of proxy containers that need to be created")
    parser.add_argument('--expected-address', default=DEF_EXPECTED_ADDR, help="expected IP address from the message-queue (incoming messages address proxies using 'http://<expected-address>:<proxy_port>')")
    parser.add_argument('--start-port', default=DEF_START_PORT, type=int, help="the starting port the assign to the proxies (every proxy will have assigned a port number +1 of the previews)")
    parser.add_argument('--proxy-file', default=DEF_PROXY_PATH, help="full path to the proxy file to be written")
    args = parser.parse_args()

    try:
        # Initialize proxies
        proxies.initiate_proxies(args.replicas, args.start_port, args.expected_address)

        # Write the proxy file at the path specified
        proxies.write_proxy_file(args.proxy_file)

        # Connecto to RabbitMQ
        rabbitmq.connect(host=args.rabbit_host, port=args.rabbit_port, user=args.rabbit_user, passwd=args.rabbit_pass)

        # Declare the exchange and queue that we are going to listen
        rabbitmq.exchenage_declare(exchange_name=args.exchange_name, durable=True)
        rabbitmq.queue_declare(queue_name=args.queue_name)

        # Binsing queue to exchange
        rabbitmq.bind_queue(exchange_name=args.exchange_name, queue_name=args.queue_name)

        # Declare the method of consumption
        rabbitmq.basic_consume(queue=args.queue_name, on_message_callback=callback, auto_ack=False)

        # Start consuming messages
        rabbitmq.start_consuming()
    except Exception as e:
        # Print the error in the console
        print(e)

        # Terminate all started proxies (if any)
        proxies.terminate_proxies()
        # Close connection with RabbitMQ
        rabbitmq.close_connection()

        # Exit with error
        sys.exit(-1)


if __name__ == "__main__":
    main()
